crates/runtime/src/traphandlers.rs

```diff
@@ -397,3 +429,120 @@ mod tls {
         })
     }
 }
+
+/// A module for registering a custom alternate signal stack (sigaltstack).
+///
+/// Rust's libstd installs an alternate stack with size `SIGSTKSZ`, which is not always
+/// large enough for our signal handling code, and which as of this writing does not
+/// have a guard page. Override it by creating and registering our own alternate stack
+/// that is large enough and has a guard page.
+#[cfg(unix)]
+mod unix_sigaltstack {
+    use std::cell::RefCell;
+    use std::convert::TryInto;
+    use std::ptr::null_mut;
+
+    thread_local! {
+        /// Thread-local state is lazy-initialized on the first time it's used, and dropped
+        /// when the thread exits.
+        pub(crate) static TLS: RefCell<Tls> = RefCell::new(unsafe { Tls::new() });
+    }
+
+    /// The size of the sigaltstack (not including the guard, which will be added).
+    /// Make this large enough to run our signal handlers.
+    static STACK_SIZE: usize = libc::SIGSTKSZ * 4;
```

Comment:
FYI in Dfinity we use std::cmp::max(SIGSTKSZ, 24 * 1024) here, because SIGSTKSZ on MacOS is already huge (128K) and reserving 512K per thread for the signal stack seemed like too much